Add ARM DesignStart FPGA V2C-DAPLink shield support

henrikbrixandersen opened the pull request:
Add support for the ARM DesignStart FPGA V2C-DAPLink shield.

henrikbrixandersen committed:
boards: arty: add support for daplink qspi mux

The ARM Cortex-M1/M3 DesignStart FPGA reference designs support booting
from an off-board, memory-mapped QSPI NOR flash device. The V2C DAPlink
shield from ARM Ltd. provides support for this feature.

If the board/shield is not configured for booting from the QSPI NOR
device, that same device is available as a regular QSPI NOR flash
device.

The presense of a shield configured for QSPI NOR flash boot is indicated
through an IRQ line used as a level-detect, non-interrupt signal.

Introduce a board specific devicetree binding for the DAPLink QSPI MUX
and provide a board specific API accessing it. Automatically set the
QSPI MUX to to provide regular QSPI NOR flash access if the board/shield
is not configured for memory-mapped QSPI NOR flash boot.

Signed-off-by: Henrik Brix Andersen <henrik@brixandersen.dk>

henrikbrixandersen committed:
boards: arm: arty: add missing default for enabling SPI support

Add missing default for enabling CONFIG_SPI if CONFIG_FLASH is
enabled. This is needed for CONFIG_SPI_NOR.

Signed-off-by: Henrik Brix Andersen <henrik@brixandersen.dk>

henrikbrixandersen committed:
boards: shields: add support for the ARM Ltd. V2C DAPLink shield

Add support for the ARM DesignStart DAPLink board for Digilent Arty
S7/A7 (V2C DAPLink) shield.

This shield is designed to be used along with the ARM Cortex-M1/M3
DesignStart FPGA reference designs running on the Digilent Arty.

Signed-off-by: Henrik Brix Andersen <henrik@brixandersen.dk>

henrikbrixandersen committed:
boards: arm: arty: add reference to the v2c_daplink shield

Add a reference to the V2C-DAPLink shield from the Digilent Arty board
documentation.

Signed-off-by: Henrik Brix Andersen <henrik@brixandersen.dk>

henrikbrixandersen commented:
Adding this to the v2.4.0 milestone in the hope we can ship this release with "full" ARM DesignStart FPGA reference design support. This shield is an official companion shield from ARM Ltd. to aid in getting started with ARM DesignStart FPGA.

Using the shield requires boot-up configuration of the QSPI MUX in some configurations, so the shield is not easily used out-of-tree without the `board.c` logic introduced here.

erwango commented on boards/arm/arty/board.c:
```diff
@@ -58,9 +58,9 @@ int board_daplink_qspi_mux_select(enum board_daplink_qspi_mux_mode mode)
-	 * shield is fitted.
+	 * The DAPLINK_fitted_n signal is routed to an IRQ line. It is used as a
+	 * level-detect non-interrupt signal to determine if the DAPLink shield
+	 * is fitted.
```
This should be moved to first commit

henrikbrixandersen replied:
@erwango Thank you for reviewing. Good catch! I must have performed a fixup on the wrong commit at some point. Fixed.

erwango commented on boards/arm/arty/dts/bindings/arm,daplink-qspi-mux.yaml:
```diff
@@ -12,7 +12,7 @@ properties:
       description: |
         IRQ line connected to the level-detect non-interrupt DAPLink shield
-        presence signal
+        fitted signal
```
Ditto

henrikbrixandersen replied:
Fixed, thanks.

erwango reviewed (changes requested):
Minor comment.
Otherwise, from shield point of view, this looks correct.

erwango reviewed (approved):
LGTM from shield side

henrikbrixandersen commented:
Rebased on master after #27888 and #24873 were merged. Refactored shield dts files a bit to allow for Cortex-M3 reference design support, added Kconfig for setting the initialization level for the board init function (due to #27985).

carlescufi commented:
@jfischer-phytec-iot could you take a look?
